keps/sig-node/5067-pod-generation/README.md

```diff
@@ -0,0 +1,931 @@
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+No risks have been identified.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+#### API server
+
+For a newly created pod, the API server will set `generation` to 1. For every 
+update to the pod that requires actuation by the kubelet, the API Server will 
+increment `generation` by 1. Pod updates that result in `generation` being 
+incremented are any changes to mutable fields in the [PodSpec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/#podspec-v1-core), and also changes to annotations
+that end up getting copied to the PodSpec such as seccomp and AppArmor annotations.
+
+Pod updates that would not result in `generation` being incremented include:
+* Changes to metadata. This means that if a Pod uses the downward API to make
+pod metadata available to containers, Pod behavior can change without the
+generation being incremented. We will consider this working as intended. 
+* Changes to status.
+
+```
+<<[UNRESOLVED]>>
+Should changes to `DeletionTimestamp` cause `generation` to be incremented? 
+<<[/UNRESOLVED]>>
+```
+
+Any attempts by clients to set or modify the `generation` field themselves will be
+ignored and overridden by the API Server. This is consistent with existing 
+behavior of the `generation` field in deployments. 
+
+#### Kubelet
+
+When the Kubelet updates the pod status as part of the [pod sync loop](https://github.com/kubernetes/kubernetes/blob/45d0fddaf1f24f7b559eb936308ce2aeb9871850/pkg/kubelet/pod_workers.go#L1214), 
+it will set the `status.observedGeneration` in the pod to reflect the pod `generation` 
+corresponding to snapshot of the pod currently being synced. That means if the 
+pod spec gets updated concurrently while the kubelet is performing a pod sync loop 
+on a previous update, the `observedGeneration` will be behind `generation`. 
+
+Outside of the pod sync loop, another place where the kubelet sets the pod status is
+when a [pod is rejected (during `HandlePodAnnotations`)](https://github.com/kubernetes/kubernetes/blob/8770bd58d04555303a3a15b30c245a58723d0f4a/pkg/kubelet/kubelet.go#L2335). This
+code will also be modified to populate `observedGeneration` to express which
+`generation` was rejected. In this case, the kubelet will not be updating the pod
+through the sync loop (due to the rejection).
```

Comment:
Just did some digging, the only other place I found is the readiness & startup probe updates (e.g. https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/status/status_manager.go#L132). I think it is correct to leave ObservedGeneration unchanged for these, as the probe that updated the status would be the one synced in the last pod sync loop (sort of moot since probes are immutable, but that could change)